### PR TITLE
Modern Authentication for Legacy Apps

### PR DESCRIPTION
**Pull request by chenkie**
screenshot

### CODE REVIEW

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+tags: 
+- authentication
+- jwt
+- javascript
+- nodejs
+---
+
+---
+
+**TL;DR:** Traditional authentication doesn't work so well for today's modern web applications for several reasons. We do, however, have a great alternative with JSON Web Tokens (JWTs) which can help us to keep our applications stateless. While cookies aren't always used to handle JWT authentication, we can make use of them in our apps if we need to. To see how this is done, take a look at the [repo](https://github.com/auth0/cookie-jwt-auth) for this tutorial.
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: I don't like the TL;DR. 

I'd mention:
* JWT makes my app be stateless like HTTP, which with cookies is hard because of session.
* If you have an old app and want to migrate it to JWTs part by part, you can use a cookie as a storage mechanism for the JWT instead of an authentication method which is what brwosers do by default.

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+---
+Authentication for web applications has historically been fairly straight forward and followed a common pattern, but this has been changing significantly in recent years. There are several reasons for these changes, and the prominent ones are related to the way modern applications are built and distributed.  In this article, we'll see how things have changed and also some ways we can adapt legacy applications to modern day practices.
+
+## Traditional Web Applications
+
+In a general sense, traditional websites and applications typically implement user authentication by using **sessions** and **cookies**. When a user logs in to a site, his or her username and password are matched against database entries. If the login is successful, the server saves the user's authentication state in memory and sends a cookie back in the reponse that contains some data, which in most cases includes the user's ID. On subsequent requests made to the server, the browser automatically sends the cookie *if* that cookie is marked as belonging to the domain to which the request is being made.
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: > the browser automatically sends the cookie *if* that cookie is marked as belonging to the domain to which the request is being made

Wording is hard. Just say that if the cookie was saved for domain auth0.com, then all requests that are sent to auth0.com automatically send that cookie. That's a browser feature that works great for traditional webapps.

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+![](https://cdn.auth0.com/blog/legacy-app-auth/legacy-app-auth-1.png)
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: Always add alts for SEO :).

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+Using APIs in this fashion is great, but things can become challenging when it comes to authentication. The traditional approach of using sessions and cookies for the user's identity doesn't work so well in these cases since their use introduces **state** to the application. One of the tenets of a RESTful API is that it should be **stateless**, meaning that when a request is made, a response within certain parameters can always be anticipated without side effect. A user's authentication state introduces such a side effect, which breaks this principle.
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: Why should they be stateless? Mention mantainability and ease of debugging :)

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+Another common pattern seen with modern web applications is that they often rely on downstream services. For example, a call to the main application server might make a request to a downstream server before the original request is resolved. The issue here is that cookies don't "flow" easily to the downstream servers and can't tell those servers about the user's authentication state.
+
+## A Modern Alternative: The JSON Web Token
+
+To combat the issues detailed above, we can take a token-based approach by using JSON Web Tokens (JWTs) for authentication. A JWT contains three parts:
+
+**1. Header**
+
+The header tells us about the algorithm and token type. It is Base64 encoded.
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: Base64URL encoded, not just base64

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+**2. Payload**
+
+The payload contains any arbitrary information in the form of claims that we as developers find useful for our applications. Some common examples include the user's ID and name. The payload is also Base64 encoded.
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: UserID is actually part of the spec and must be sent as `sub`

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+**3. Signature**
+
+The signature is based on a secret key that only the server knows and is used to verify the authenticity of the token.
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: Mention this can use lots of algorithms, some that use one shared secret (HMAC) or some that use public private key secrets (RSA)

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+![](https://cdn.auth0.com/blog/legacy-app-auth/legacy-app-auth-5.png)
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: Alt

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+To use a JWT for authentication, a token must first be received from the server. The user enters his or her credentials as per usual, but instead of creating a session and returning a cookie, the server will respond with a JWT. The JWT then needs to be saved locally, which is normally done with local storage. However, as we'll see in the next section, it is possible to save the JWT in a cookie.
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: > To use a JWT for authentication, a token must first be received from the server

Rephrase it's not clear

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+The JWT must be sent to the server to access protected routes, and it is typically sent as an `Authorization` header. The scheme used for this header is `Bearer`, and thus the full header looks like this:
+
+```js
+Authorization: Bearer <token>
+```
+
+Middleware on the protected API routes will check for a valid JWT, and if there is one, will let the request through and return the data being requested.
+
+It should be reiterated that at no point is the user's state saved in memory on the server, meaning that the user isn't "logged in" in the conventional sense. However, a valid JWT gives the user the keys to access data each time a request is made, and in this way, a stateless authentication mechanism is in place.
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: Mention that the API has all of the user information because it's embed inside the token. So not only it's stateless, but also, it has less altency since it doesn't need to go to the DB to get user information, it's just there ;)

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+![](https://cdn.auth0.com/blog/legacy-app-auth/legacy-app-auth-2.png)
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: alt

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+
+Using a JWT for authentication helps to solve the challenges noted above. We can fully rely on data APIs that are stateless and even make requests to downstream services. It also doesn't matter which domain is serving our API, nor does it matter which specific server a request goes to if the app is deployed across many.
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: Why can we make requests to downstream services? Because JWT is a spec for authentication that all services provide. Link to show all the frameworks from jwt.io.

Also mention that cookies each server uses their own scheme which actually makes it hard to connect to downstrea services. Not sure if this is here or in the downstream point above :). your call.

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+
+## Bridging the Gap: Using JWTs in Cookies
+
+Modernizing legacy applications by implementing token-based authentication can lead to many gains. If it isn't feasible for an organization to totally throw away cookies, there are still ways to make token authentication viable. Let's take a look at how this can be done with NodeJS and the **express-jwt** package.
+
+To follow along, you can download Auth0's [NodeJS seed project](https://auth0.com/docs/node-auth0/master/create-package?path=examples/nodejs-api&filePath=&type=server) from the [docs page](https://auth0.com/docs/quickstart/backend/nodejs/).
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: Mention that even though it's an Auth0 seed project, this works with absolutely any service. It just checks JWTs ;)

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+  app.use(cookieParser());
+
+  ...
+
+});
+```
+
+### Authenticating with a Cookie
+
+We'll need to retrieve a JWT for a user and save it locally as a cookie. To find out more about retrieving JWTs for users in your Auth0 account, read the [API documentation](https://auth0.com/docs/api/v2).
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: If you're using Auth0, then ..... Remember this is a generic blogpost where we link things that show Auth0, but it's not fully Auth0 related.

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+...
+
+var authenticate = jwt({
+  secret: new Buffer(process.env.AUTH0_CLIENT_SECRET, 'base64'),
+  audience: process.env.AUTH0_CLIENT_ID,
+  // Custom function to retrieve the JWT
+  // We first look for the JWT in a header and if it isn't there,
+  // we look for it in a cookie
+  getToken: function fromHeaderOrCookie(req) {
+    if(req.headers.authorization && req.headers.authorization.split(' ')[0] === 'Bearer') {
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: Nice!

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+
+We've added in the `getToken` function as an option on the object that is passed into `jwt`. Our implementation here says that we want to first look for the presence of a JWT as an `Authorization` header, and if there isn't anything there, we want to check whether there is a cookie with a name of `id_token`.
+
+With the `id_token` cookie in place, we can see that requests to the `secured/ping` endpoint go through just fine.
+
+![](https://cdn.auth0.com/blog/legacy-app-auth/legacy-app-auth-3.png)
+
+## Aside: JWT Authentication is Easy with Auth0
+
+Auth0 issues JSON Web Tokens on every login for your users. That means that you can have a solid identity infrastructure, including Single Sign On, User Management, support for Social (Facebook, Github, Twitter, etc.), Enterprise (Active Directory, LDAP, SAML, etc.) and your own database of users with just a few lines of code.
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: Rephrase this aside and create another one.

Mention that Auth0 uses JWT for everything and as you saw in this post, it's really easy to switch from regular authentication to JWT with a lot of things to gain. If you don't want to implement this change on your own, just use Auth0 :). And ask @iamevilrabbit for a picture of Auth0 badge and a JWT logo getting together nice

**Comment by mgonto** [comment]
Ask @iamevilrabbit to redo the the sequence diagram becase black doesn't look good with the post

File: _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+  avatar: https://www.gravatar.com/avatar/7f4ec37467f2f7db6fffc7b4d2cc8dc2?size=200
+design: 
+  bg_color: "#842946"
+  image: https://cdn.auth0.com/blog/jwt/logo-400.png
+  image_size: "170%"
+tags: 
+- authentication
+- jwt
+- javascript
+- nodejs
```
Comment (mgonto) [inline _posts/2015-09-25-modern-authentication-for-legacy-apps.markdown]: add modern authentication, legacy apps, cookie authentication as tags

**Comment by chenkie** [comment]
SEO Checklist: https://docs.google.com/a/auth10.com/document/d/1SkUEqRHdVr0xn87GqNfbYG39Q6Y6xLPdKpDoSGEg0fo/edit?usp=sharing
